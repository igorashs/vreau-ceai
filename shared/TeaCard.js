import styled from 'styled-components';
import Image from 'next/image';
import { StyledLink } from '@/shared/StyledLink';
import breakpoints from 'GlobalStyle/breakpoints';

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  gap: calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 1px solid var(--accent-dark);
  padding: calc(var(--baseline) / 2);
  background-color: #fff;

  @media (min-width: ${breakpoints.lg}) {
    display: grid;
    grid-template-columns: auto 1fr;
  }
`;

const ImgContainer = styled.figure`
  width: 224px;
  height: 224px;
  overflow: hidden;
  border-radius: 4px;
  border: 1px solid var(--accent-text-light);
`;

const Title = styled.h5`
  grid-column: 1;
  text-align: center;
`;

const Description = styled.p`
  align-self: center;
  grid-row: 1;
  grid-column: 2;
  overflow: hidden;
  text-overflow: ellipsis;
  display: -webkit-box;
  -webkit-line-clamp: 3;
  -webkit-box-orient: vertical;

  @media (min-width: ${breakpoints.lg}) {
    -webkit-line-clamp: 4;
  }
`;

const Price = styled.p`
  text-align: center;
  grid-column: 1;
  font-weight: 500;
`;

const Actions = styled.div`
  display: grid;

  @media (min-width: ${breakpoints.lg}) {
    display: flex;
    justify-content: flex-end;
    gap: var(--baseline);
  }
`;

export const TeaCard = ({ tea, category }) => {
  return (
    <Wrapper>
      <ImgContainer>
        <Image
          width={700}
          height={700}
          src={`/uploads/${tea.src}`}
          alt={tea.name}
        />
      </ImgContainer>
      <Title>{tea.name}</Title>
      <Description>{tea.description}</Description>
      <Price>{`${tea.price}lei - ${tea.quantity}g`}</Price>
      <Actions>
        <StyledLink
          href={`/categories/${category}/${tea.name}`}
          button
          text="vezi produs"
        />
      </Actions>
    </Wrapper>
  );
};
