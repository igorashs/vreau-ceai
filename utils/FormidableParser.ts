import * as validator from '@/utils/validator';
import FormidableNameSpace, { Files, File, IncomingForm } from 'formidable';
import { NextApiRequest } from 'next';
import { Formidable, ProductFields } from 'types';
import { productMessages } from './validator/schemas/product';
import { NextApiSessionRequest } from './withSession';

type ParsedData = {
  fields: Partial<ProductFields>;
  files: Files & { src?: File };
};

type HandleProductForm = (
  req: NextApiRequest | NextApiSessionRequest,
  options: Partial<FormidableNameSpace.Options>,
  allowedFileTypes: string[],
) => Promise<ParsedData>;

const handleProductForm: HandleProductForm = async (
  req,
  options,
  allowedFileTypes,
) => {
  const data: ParsedData = await new Promise((resolve, reject) => {
    const form: Formidable = new IncomingForm(options);

    form.onPart = (part) => {
      if (part.filename) {
        // get file extension
        const fileType = part.mime?.split('/').pop() ?? '';

        // validate allowed file types
        if (!allowedFileTypes.includes(fileType)) {
          form.emit(
            'error',
            validator.createValidationError({
              message: productMessages.src.invalid,
              key: 'src',
            }),
          );
        }
      }

      form.handlePart(part);
    };

    form.parse(req, (err, fields, files) => {
      if (err) {
        // custom error on maxFileSize
        if (err.message.includes('maxFileSize')) {
          reject(
            validator.createValidationError({
              message: productMessages.src.max,
              key: 'src',
            }),
          );
        } else {
          reject(err);
        }
      }

      resolve({ fields, files });
    });
  });

  return data;
};

export default {
  handleProductForm,
};
