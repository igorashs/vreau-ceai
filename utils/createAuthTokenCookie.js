import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createAuthTokenCookie = (payload) => {
  const authToken = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  return cookie.serialize('auth-token', authToken, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });
};
