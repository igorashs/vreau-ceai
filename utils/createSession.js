import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createSession = (payload) => {
  const token = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  const [header, jwtPayload, signature] = token.split('.');
  const session = header + '.' + jwtPayload;

  const sessionCookie = cookie.serialize('session', session, {
    secure: process.env.NODE_ENV !== 'development',
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  const authCookie = cookie.serialize('auth', signature, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  return [sessionCookie, authCookie];
};
