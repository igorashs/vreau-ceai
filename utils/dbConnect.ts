import mongoose from 'mongoose';

const dbConnect = async () => {
  if (mongoose.connection.readyState >= 1) {
    return;
  }

  await mongoose.connect(process.env.MONGODB_URI as string, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
    useFindAndModify: false,
  });
};

export default dbConnect;
