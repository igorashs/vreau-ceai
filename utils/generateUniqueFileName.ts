import { nanoid } from 'nanoid';

const generateUniqueFileName = (fileName: string) => {
  // grab file extension
  const matches = fileName?.match(/\..+$/);

  // create new file name
  const uniqueFileName = nanoid() + (matches && matches[0]);

  return uniqueFileName;
};

export default generateUniqueFileName;
