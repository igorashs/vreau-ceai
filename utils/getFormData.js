export const getFormData = (data) => {
  const formData = new FormData();

  for (const name in data) {
    if (name === 'src') continue;

    formData.append(name, data[name]);
  }

  if (data.src[0]) {
    formData.append('src', data.src[0]);
  }

  return formData;
};
