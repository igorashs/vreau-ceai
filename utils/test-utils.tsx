import { render, RenderOptions } from '@testing-library/react';
import GlobalStyle from 'GlobalStyle';

const Wrapper: React.FC = ({ children }) => (
  <>
    <GlobalStyle />
    {children}
  </>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';

export { customRender as render };
