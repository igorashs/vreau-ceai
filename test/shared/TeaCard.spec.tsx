import { render } from '@/utils/test-utils';
import TeaCard from '@/shared/TeaCard';
import { ImageProps } from 'next/image';

jest.mock('next/image', () => ({ alt, ...all }: ImageProps) => (
  <img {...all} alt={alt} />
));

describe('TeaCard', () => {
  it('renders correctly', () => {
    const tea = {
      _id: '1',
      name: 'green-tea',
      src: 'tea.png',
      description: 'tasty tea',
      price: 69,
      quantity: 100,
      total_quantity: 500,
      category_id: '1',
      recommend: false,
    };

    const category = 'green';

    const { container, getByAltText, getByText, getByRole } = render(
      <TeaCard tea={tea} category={category} />,
    );

    expect(getByAltText(tea.name)).toHaveAttribute(
      'src',
      expect.stringContaining(tea.src),
    );

    expect(getByText(tea.name)).toBeVisible();
    expect(getByText(tea.description)).toBeVisible();
    expect(getByText(tea.price, { exact: false })).toBeVisible();
    expect(getByText(tea.quantity, { exact: false })).toBeVisible();
    expect(getByRole('link')).toHaveAttribute(
      'href',
      expect.stringContaining(`${category}/${tea.name}`),
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
