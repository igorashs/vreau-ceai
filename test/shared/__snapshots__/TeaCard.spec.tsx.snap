// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeaCard renders correctly 1`] = `
.c6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 7px;
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c6:hover {
  background-color: var(--accent);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 1px solid var(--accent-dark);
  padding: calc(var(--baseline) / 2);
  background-color: #fff;
}

.c1 {
  width: 224px;
  height: 224px;
  overflow: hidden;
  border-radius: 4px;
  border: 1px solid var(--accent-text-light);
}

.c2 {
  grid-column: 1;
  text-align: center;
}

.c3 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  grid-row: 1;
  grid-column: 2;
  overflow: hidden;
  text-overflow: ellipsis;
  display: -webkit-box;
  -webkit-line-clamp: 3;
  -webkit-box-orient: vertical;
}

.c4 {
  text-align: center;
  grid-column: 1;
  font-weight: 500;
}

.c5 {
  display: grid;
}

@media (min-width:992px) {
  .c0 {
    display: grid;
    grid-template-columns: auto 1fr;
  }
}

@media (min-width:992px) {
  .c3 {
    -webkit-line-clamp: 4;
  }
}

@media (min-width:992px) {
  .c5 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: end;
    -webkit-justify-content: flex-end;
    -ms-flex-pack: end;
    justify-content: flex-end;
    gap: var(--baseline);
  }
}

<div
  class="c0"
>
  <figure
    class="c1"
  >
    <img
      alt="green-tea"
      height="700"
      src="/uploads/tea.png"
      width="700"
    />
  </figure>
  <h5
    class="c2"
  >
    green-tea
  </h5>
  <p
    class="c3"
  >
    tasty tea
  </p>
  <p
    class="c4"
  >
    69lei - 100g
  </p>
  <div
    class="c5"
  >
    <a
      aria-label="vezi produs"
      class="c6"
      href="/categories/green/green-tea"
      rel=""
      target=""
    >
      vezi produs
    </a>
  </div>
</div>
`;
